crates/uv-workspace/src/workspace.rs

```diff
@@ -656,57 +678,113 @@ impl Workspace {
         workspace_definition: ToolUvWorkspace,
         workspace_pyproject_toml: PyProjectToml,
         current_project: Option<WorkspaceMember>,
-        options: &DiscoveryOptions<'_>,
+        options: &DiscoveryOptions,
+        cache: &WorkspaceCache,
     ) -> Result<Workspace, WorkspaceError> {
-        let mut workspace_members = BTreeMap::new();
-        // Avoid reading a `pyproject.toml` more than once.
-        let mut seen = FxHashSet::default();
-
-        // Add the project at the workspace root, if it exists and if it's distinct from the current
-        // project.
-        if current_project
-            .as_ref()
-            .map(|root_member| root_member.root != workspace_root)
-            .unwrap_or(true)
-        {
-            if let Some(project) = &workspace_pyproject_toml.project {
-                let pyproject_path = workspace_root.join("pyproject.toml");
-                let contents = fs_err::read_to_string(&pyproject_path)?;
-                let pyproject_toml = PyProjectToml::from_string(contents)
-                    .map_err(|err| WorkspaceError::Toml(pyproject_path.clone(), Box::new(err)))?;
+        let cache_key = WorkspaceCacheKey {
+            workspace_root: workspace_root.clone(),
+            discovery_options: options.clone(),
+        };
```

Comment:
We could avoid the clone here with a nested hashmap but I prefer the simplicity and extensibility (with future discovery options) here.